### ISSUE TITLE
Safari Fairplay MediaKeys Polyfill 

### ISSUE THREAD

**Issue by sanbornhilland** [google/shaka-player#382]
In issue #279 it was mentioned that "The status of Fairplay through EME is unknown at this time." Can you elaborate on what you mean by this? We have some apple sample code that shows Fairplay playback in Safari. However trying to play the same content through Shaka with an HLS Parser plugin is not going so smoothly. I'm not sure the MediaKeys polyfill is working correctly on Safari. Have you tested in Safari? 

I can get more detailed logs and specific failings in the next few days but I wanted to initially enquire about the compatibility of the MediaKeys polyfill and the Safari (9.1) EME implementation.

**Comment by joeyparrish** [comment]
What I meant by that comment is that we have never seen any Fairplay-compatible DASH content we could try in Shaka, nor any EME-based Fairplay demos using HLS.  So we can only assume Apple's EME works.  Can you share a link to something of that sort?

A quick check shows that Safari 9's EME is the webkit-prefixed v0.1b API.  The polyfill for that was originally developed against Chrome, but Chrome has long-since implemented the unprefixed working-draft EME API.  If we need to modify the polyfill to better support Safari, I have no problems with that.  Safari is, at the moment, the only known consumer of it.

**Comment by sanbornhnewyyz** [comment]
We have a server issue to resolve but then I should be able to provide you with a working Safari sample with HLS Fairplay content.

**Comment by sanbornhnewyyz** [comment]
I'm still working on getting you some HLS + Fairplay samples to give you. But there is sample client code from apple here: https://developer.apple.com/streaming/fps/ if you download the FairPlay Streaming Server SDK. If you haven't looked at this it would be useful for comparing to the v01b polyfill.

For instance, for FPS on Safari the init data is formed by concatenating the `initData` from the `needkey` event with a `contentId` and an Apple App Cert. I believe this is different from Chrome and other browsers that just require the init data? It looks to me like currently Shaka and the MediaKeys polyfill assumes the `needkey` `event.initData` is sufficient.

After that it looks like there are some other differences. The concatenated init data is passed in the `createSession` call and then there is no `generateRequest` call at all. This also appears to be different than how Shaka is handling it where the `createSession` call does not take any arguments and there is a `generateRequest` call that takes `initData`.

It appears to me that the flow on Safari is a bit different than on Chrome. I have tested that Apple sample code with HLS + Fairplay content and it does work in Safari.

I'm also having the following issue in Safari: In `shaka.polyfill.PatchedMediaKeys.v01b.MediaKeys.prototype.onWebkitNeedKey_` it throws an error when trying to dispatch the shaka `FakeEvent`: `UNSPECIFIED_EVENT_TYPE_ERR: DOM Events Exception 0: The Event's type was not specified by initializing the event before the method was called.` The only way I can see to avoid that is to create a custom event instead like this:

```
var event2 = new CustomEvent('encrypted');
event2.initDataType = 'webm';
event2.initData = event.initData;
```

I'm not sure you'd want to handle it like this just for consistency reasons but this was how I could get around that error just for POC purposes.

**Comment by sarge** [comment]
@joeyparrish would it help if I provide some test content to run against?

**Comment by joeyparrish** [comment]
@sarge, it couldn't hurt!  Thanks!

**Comment by boredom2** [comment]
Hi there - we are now also finally under Pressure to run a DRM Solution for Safari asap.
I am really a little confused :) We have Assets in HLS and with FairPlay Protection ready.
Initially, I thought, that Safari will handle the Process itself, but looks, like if we have to do something here :) So Apple DRM also works on MSE Base?
And if yes, I suggest, this will only work with HLS (DASH with FairPlay is afaik not possible to create or at least never done and for sure not possible to create on MS Azure).
So can you maybe give me a little Hint on how to make this work?
Did someone ever made HLS+FairPlay+MSE work at all?
(and if you know that - how does this work on iOS, if this does not support MSE?)

**Comment by sarge** [comment]
@joeyparrish I have uploaded a working sample of the fairplay drm. 
https://s3.amazonaws.com/shift72-temp/hls_fps_bento4_sintel/master.m3u8

This file contains reference to the "Apple Developer Program License Agreement", so you can choose not to view it. Not sure what to do about worlds colliding here (shrug)
https://s3.amazonaws.com/shift72-temp/hls_fps_bento4_sintel/safari_hls.html

@boredom2 HLS + Fairplay works on Safari Desktop. But does not work on Safari iOS, an app is still required.
Dash + Fairplay + (Widevine + Playready) + MSE appears possible, implementations are rare from what I have seen. An additional pssh box with the fairplay id and some work generating the initData in the format  [initData] = [initData] + [4 byte: idLength] + [idLength byte: id] + [4 byte:certLength] + [certLength byte: cert] should work.

**Comment by boredom2** [comment]
Thanks for that Information - I will try once SHAKA can handle the slightly "crippled" HLS Manifest, that we get from MS Azure CDN. I will update then here :)

**Comment by joeyparrish** [comment]
Here's an update on our progress.

The polyfill part is not that difficult.  I have written a polyfill that builds a compliant MediaKeys API on top of Safari 10's WebKitMediaKeys.

Parsing the FairPlay tag in HLS to feed EME is trivial.

The challenge is that the key system `com.apple.fps.1_0` seems to be usable only with `video.src=foo.m3u8`.  When used with MediaSource, `createSession()` causes the page to crash and reload.  And Shaka Player currently only works with MediaSource.

Judging from a [WebKit commit log](https://trac.webkit.org/changeset/164529/webkit), they introduced `com.apple.fps.2_0` for MediaSource:
 > (WebCore::keySystemIsSupported): Use "com.apple.fps.2_0" to distinguish from the not-media-source scheme.

I have not been able to find any working examples of that key system in use.

Creating a session with `fps.2_0` results in a message that contains only the text "certificate".  Looking at WebKit sources, [this is unconditional](https://github.com/WebKit/webkit/blob/a9461060da0db60d0c0493ebf599c40273ba1b29/Source/WebCore/platform/graphics/avfoundation/objc/CDMSessionAVStreamSession.mm#L107) and we are [expected to call `update()` with the certificate](https://github.com/WebKit/webkit/blob/a9461060da0db60d0c0493ebf599c40273ba1b29/Source/WebCore/platform/graphics/avfoundation/objc/CDMSessionAVStreamSession.mm#L185) after that.

Dealing with this certificate request is enough to do, and we can even hide it in the MediaKeys polyfill.  However, the message we get after that is not a license request, either.  It's completely empty.

WebKit sources also [show support for `com.apple.fps.3_0`](https://github.com/WebKit/webkit/blob/fe0d7c6f1d662f675addbca632b248cc58734a94/Source/WebCore/platform/graphics/avfoundation/CDMPrivateMediaSourceAVFObjC.mm#L47), but that also has no documentation and does not get us any further than `fps.2_0` did.

**Comment by sarge** [comment]
Thanks Joey for taking a look at this. You got further that I did. Would you mind pushing a branch?

```fps.2_0``` being for MSE is a good find - the apple forums note that NetFlix uses this scheme.

When you say after the certificate request you get a message that is empty do you mean that get a call to ```webkitkeymessage``` with no initdata?

**Comment by joeyparrish** [comment]
I mean that the second `webkitkeymessage` event has `message.length` of zero.

**Comment by joeyparrish** [comment]
I can't push anything until it's been through code review by the team, but I'll try to clean up what I've done so far and make it available in some form.

**Comment by avelad** [comment]
Safari 11, now it's in beta with macOS High Sierra. Are there any change on eme-fairplay?

Shaka Player support.html reports:

> Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13) AppleWebKit/604.1.21 (KHTML, like Gecko) Version/11.0 Safari/604.1.21
> v2.1.3
> 
> {
>   "manifest": {
>     "application/dash+xml": true,
>     "application/x-mpegurl": true,
>     "application/vnd.apple.mpegurl": true,
>     "application/x-offline-manifest": true,
>     "mpd": true,
>     "m3u8": true,
>     "application/vnd.ms-sstr+xml": false,
>     "ism": false
>   },
>   "media": {
>     "video/mp4; codecs=\"avc1.42E01E\"": true,
>     "video/mp4": true,
>     "video/mp4; codecs=\"avc3.42E01E\"": true,
>     "video/mp4; codecs=\"hvc1.1.6.L93.90\"": true,
>     "audio/mp4; codecs=\"mp4a.40.2\"": true,
>     "audio/mp4": true,
>     "audio/mp4; codecs=\"ac-3\"": true,
>     "audio/mp4; codecs=\"ec-3\"": true,
>     "video/webm; codecs=\"vp8\"": false,
>     "video/webm": false,
>     "video/webm; codecs=\"vp9\"": false,
>     "video/webm; codecs=\"av1\"": false,
>     "audio/webm; codecs=\"vorbis\"": false,
>     "audio/webm": false,
>     "audio/webm; codecs=\"opus\"": false,
>     "video/mp2t; codecs=\"avc1.42E01E\"": true,
>     "video/mp2t": true,
>     "video/mp2t; codecs=\"avc3.42E01E\"": true,
>     "video/mp2t; codecs=\"hvc1.1.6.L93.90\"": true,
>     "video/mp2t; codecs=\"mp4a.40.2\"": true,
>     "video/mp2t; codecs=\"ac-3\"": true,
>     "video/mp2t; codecs=\"ec-3\"": true,
>     "text/vtt": true,
>     "application/mp4; codecs=\"wvtt\"": true,
>     "application/mp4": true,
>     "application/ttml+xml": true,
>     "application/mp4; codecs=\"stpp\"": true
>   },
>   "drm": {
>     "org.w3.clearkey": null,
>     "com.widevine.alpha": null,
>     "com.microsoft.playready": null,
>     "com.apple.fps.2_0": null,
>     "com.apple.fps.1_0": null,
>     "com.apple.fps": null,
>     "com.adobe.primetime": null
>   },
>   "offline": true
> }

In MAC OSX El capitan reports:

> Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/601.7.7 (KHTML, like Gecko) Version/9.1.2 Safari/601.7.7
> v2.1.3
> 
> 
> {
>   "manifest": {
>     "application/dash+xml": true,
>     "application/x-mpegurl": true,
>     "application/vnd.apple.mpegurl": true,
>     "application/x-offline-manifest": true,
>     "mpd": true,
>     "m3u8": true,
>     "application/vnd.ms-sstr+xml": false,
>     "ism": false
>   },
>   "media": {
>     "video/mp4; codecs=\"avc1.42E01E\"": true,
>     "video/mp4": true,
>     "video/mp4; codecs=\"avc3.42E01E\"": false,
>     "video/mp4; codecs=\"hvc1.1.6.L93.90\"": false,
>     "audio/mp4; codecs=\"mp4a.40.2\"": true,
>     "audio/mp4": true,
>     "audio/mp4; codecs=\"ac-3\"": true,
>     "audio/mp4; codecs=\"ec-3\"": true,
>     "video/webm; codecs=\"vp8\"": false,
>     "video/webm": false,
>     "video/webm; codecs=\"vp9\"": false,
>     "video/webm; codecs=\"av1\"": false,
>     "audio/webm; codecs=\"vorbis\"": false,
>     "audio/webm": false,
>     "audio/webm; codecs=\"opus\"": false,
>     "video/mp2t; codecs=\"avc1.42E01E\"": true,
>     "video/mp2t": true,
>     "video/mp2t; codecs=\"avc3.42E01E\"": false,
>     "video/mp2t; codecs=\"hvc1.1.6.L93.90\"": false,
>     "video/mp2t; codecs=\"mp4a.40.2\"": true,
>     "video/mp2t; codecs=\"ac-3\"": true,
>     "video/mp2t; codecs=\"ec-3\"": true,
>     "text/vtt": true,
>     "application/mp4; codecs=\"wvtt\"": true,
>     "application/mp4": true,
>     "application/ttml+xml": true,
>     "application/mp4; codecs=\"stpp\"": true
>   },
>   "drm": {
>     "com.apple.fps.1_0": {
>       "persistentState": true
>     },
>     "com.apple.fps": {
>       "persistentState": true
>     },
>     "org.w3.clearkey": null,
>     "com.widevine.alpha": null,
>     "com.microsoft.playready": null,
>     "com.apple.fps.2_0": null,
>     "com.adobe.primetime": null
>   },
>   "offline": true
> }

**Comment by joeyparrish** [comment]
At a glance, the EME API has not changed much in Safari 11 Tech Preview.  I have not yet investigated deeper to see if FairPlay is any easier to use.

**Comment by sarge** [comment]
I have had a dig into how other well know players appear to work. These are the functions and events I have seen. I am a little unclear on the role of ```webkitkeymessage``` at this stage.

```
[Info] EME LOGGER: WebKitMediaKeys.isTypeSupported (2) (inject.js, line 54)
"com.apple.fps.3_0"
"video/mp4"

// get the init data from the content
[Log] video.webkitneedkey called – Uint8Array (253) (inject.js, line 85)
Uint8Array (253)

// create the session with the initdata 
[Info] EME LOGGER: WebKitMediaKeys.create session (2) (inject.js, line 69)
"video/mp4"
Uint8Array (251)

// 
[Log] session.webkitkeymessage called – Uint8Array (11) (inject.js, line 73)
Uint8Array (11) 

// certificate being added
[Info] EME LOGGER: session.update called – Uint8Array (1246) (inject.js, line 61)
Uint8Array (1246)
[Info] EME LOGGER: session.update – "MIIE2jCCA8KgAwIBAgIIBRGnbPd8z1YwDQYJKoZIhvcNAQEFBQAwfzELMAkGA1UEBhMCVVMxEzARBgNVBAoMCkFwcGxlIEluYy4xJjAkBgNVBAsMHUFwcGxlIENlcnRpZmljYXRpb24g…" (inject.js, line 62)

// license being added?
[Log] session.webkitkeymessage called – Uint8Array (6576) (inject.js, line 73)
Uint8Array (6576)
[Info] EME LOGGER: sessionupdate – Uint8Array (1052) (inject.js, line 61)
Uint8Array (1052)
[Info] EME LOGGER: session.update – "AAAAAQAAAAAeuevhIffYAy83tkB6vtdqAAAEAPX5+hQW6VnoZjC6GVznn3zbfn86yM1UM22rXjEfriDHzw9lDiGPzGLYGFtMB6nT1kAwYeSvNZJ9CHxIqTZWf4VTt3Q1AGTZkgieM7Rs…" (inject.js, line 62)

// playback succeeds
```

**Comment by joeyparrish** [comment]
That looks a lot like what I was able to achieve, except that when I get the 11-byte certificate message and pass my certificate back to update(), the next message I get is empty.  I'm not sure if my cert is bad or if the original init data was to blame.  (Presumably the real license request based on init data was deferred during cert request, so it could be the init data's fault.)

**Comment by sarge** [comment]
So I reached out to Apple for support on this issue and the feedback is that this is not currently part of their public api, so would not offer any support in this matter. I filed a feature request as suggested.
It could be any number of issues at this point, I would not rule out a whitelist on the certificate or signing authority restricting this feature.

**Comment by joeyparrish** [comment]
In that case, there is probably nothing that we can do to complete this work.  If Apple actively does not want us to do it, there may be no point in continuing.  I will still endeavor to clean up and publish my work-in-progress FairPlay polyfill, in case it is useful to anyone in the future.

@sarge, is your support conversation with Apple public?  If so, could you provide a link to help document the current situation?

**Comment by sarge** [comment]
@joeyparrish Unfortunately all done comms were over email, also the feature request does not have an external link either. I will update this thread if I hear back.

I agree that this issue can be closed until we hear more.

Thanks for your efforts:)

**Comment by joeyparrish** [comment]
Hi all,

I just published a new branch called [`fairplay`](https://github.com/google/shaka-player/tree/fairplay) including just [one unique commit](https://github.com/google/shaka-player/commit/20f5751351987892fb9f9ddaa5b75c13dbdf477a) with what I was able to accomplish for FairPlay in Safari with MediaSource.

I hope this is a useful starting point for someone to help us figure out the missing pieces, assuming it is possible at all.  For now, I am moving this issue back to the backlog.  Please let us know if you have any insights or if you can contribute.

Thanks,
Joey

**Comment by chrisfillmore** [comment]
I was on a call today with Apple where they assured me HLS-FP playback was possible using MSE and EME in Safari, and that solutions existed in market which used these API's. Hopefully we can get some support from them, so I'll send you an email.

**Comment by bbert** [comment]
Hi,
Is it possible to share information you got from Apple to make HLS+FP playback working?
Guidelines from WWDC15's presentation (https://developer.apple.com/videos/play/wwdc2015/502/) seems not to be enough. Especially if a certificate is required.

**Comment by chrisfillmore** [comment]
@joeparrish I’m in touch with Apple about getting a license challenge message in Safari using MSE. You mention calling MediaKeySession#update with the certificate after you receive the phony “certificate” keymessage. I’m working on a demo to send to Apple but I’m not seeing the same behaviour you are, the empty keymessage event. I just get a webkitkeyerror from the session.

Have I understood your post correctly? I’ll take a closer look at your fairplay branch to see if I’ve missed anything. I’ve also sent you the demo app I’m working on so far, in case you want to look at it (you should receive an invite).

**Comment by chrisfillmore** [comment]
Just to add more detail, this is the callback I have attached to `webkitkeymessage`:

```
function onKeyMessage (e) {
  console.debug('webkitkeymessage was fired', e);
  var strMessage = String.fromCharCode.apply(undefined, e.message);

  if (strMessage === 'certificate') {
    console.debug('cdm requested the certificate');
    keySession.update(new Uint8Array(certificate)); // Do I call update like this??
  } else {
    console.debug('license challenge available');
  }
}
```

**Comment by chrisfillmore** [comment]
@joeyparrish I've learned it's not necessary to concatenate the initData with the certificate when using `com.apple.fps.2_0`. The initData should be passed to `createSession()` as-is.

**Comment by avelad** [comment]
@chrisfillmore with this change, can you use `com.apple.fps.2_0` without problems?

**Comment by chrisfillmore** [comment]
@avelad I can't comment on playback, because the test app I wrote doesn't get that far. But I can produce a Fairplay license challenge.

**Comment by sandeep-mirpuri** [comment]
@joeyparrish @sarge any updates on this?

**Comment by joeyparrish** [comment]
Hi all,

We haven't had time to look at this in the last couple months.  It will be a priority for v2.3, which we will be starting soon.  I apologize for the delay.

**Comment by avelad** [comment]
Hi @joeyparrish , can you schedule it in v2.3 milestone?

**Comment by avelad** [comment]
@joeyparrish any updates on this?

**Comment by chrisfillmore** [comment]
I wanted to add: I don't think it's necessary to call `DrmEngine#formatFairPlayRequest_`. Right now we're doing HLS-FP playback in Safari via `src=` and I don't do any manipulation on the message from the CDM, I pass it on directly to the server. This may depend on how the license server is configured?

**Comment by joeyparrish** [comment]
Sorry for the lack of updates.  We haven't had time to work on this lately, but we hope to get back to it soon.  This may be obviated by #997, at which point we don't need to do quite so much work to make FairPlay + MediaSource work.

**Comment by chrisfillmore** [comment]
Safari 11 is out:
https://developer.apple.com/library/content/releasenotes/General/WhatsNewInSafari/Safari_11_0/Safari_11_0.html

I notice that `requestMediaKeySystemAccess` is once again undefined. :(

**Comment by kidmiracleman** [comment]
@chrisfillmore Any reason why you need to use  com.apple.fps.2_0 rather than  com.apple.fps.1_0.

The Apple FairPlay example app uses the latter keysystem - however I do notice that most of the functionality seems to be embedded underneath the video tag (Not exposed via MSE)

**Comment by chrisfillmore** [comment]
`com.apple.fps.1_0` is for streaming via `videoElement.src = playlist.m3u8`. This is the method of playback documented in the Fairplay SDK demo.

`com.apple.fps.2_0` is for playback via MSE.

**Comment by barbarosalp** [comment]
Hi @joeyparrish,
Could you guys have a time to work on this lately?
Thanks.

**Comment by chrisfillmore** [comment]
@joeyparrish I am tinkering with the `fairplay` branch to see if I can get it working. Do you want to assign to me? It won't get done overnight, but I do have time to work on it.

**Comment by barbarosalp** [comment]
Hi @chrisfillmore, there is an Apple provided code example for fairplay. If you don't have it, I can send them.

**Comment by chrisfillmore** [comment]
Thanks @barbarosalp, I have it.

**Comment by chrisfillmore** [comment]
@joeyparrish Could someone merge update `fairplay` with the latest from `master`?

I am hoping I can submit a PR against `fairplay` with this commit:

https://github.com/chrisfillmore/shaka-player/commit/8a057d7a9b7491dc79625923dd413c396b028af2

I haven't achieved playback yet but I can make a successful license request. I'm trying to play some TS content but it's not working, I'll need to do more investigation. I'm hoping in the meantime we can review the changes I've made.

**Comment by karlg-arris** [comment]
@chrisfillmore I took the liberty of briefly reviewing this PR.
The only comment I have (for now) is that there seems to be some application specific logic embedded  within the generic library implementation.
This is concerning license acquisition from a remote KSM and subsequent response parsing. For other DRM systems the application developer is simply handed off raw bytes by Shaka, which you can then chose to prepare yourself via a network 'hook'  - e.g. the developer is free to chose their own request format (URI encode, HTTP header..etc)

The same applies for response parsing  - Shaka expects an ArrayBuffer to be returned, but the application is free to use a custom response format (and add the appropriate parsing hook)

**Comment by chrisfillmore** [comment]
Hi @karlg-arris, this implementation would still allow custom license requests and responses via the networking engine. Nothing has changed in that layer.

I think the only app-specific code which I neglected to remove was in drm_engine:

```
request.headers['Content-Type'] = 'application/octet-stream';
```

**Comment by karlg-arris** [comment]
Sorry @chrisfillmore  - thanks for clarifying !

**Comment by chrisfillmore** [comment]
Beep beep Shaka team

I'd like to submit a PR against `fairplay` so that we can advance the state of FP support in Shaka, even if it's not all the way to playback. Can you advise on next steps @joeyparrish ?

**Comment by joeyparrish** [comment]
Based on feedback from our survey, this is at the top of our list for v2.4.  I'll try to get the branch updated against master soon.

**Comment by joeyparrish** [comment]
I pushed updates to the fairplay branch.  I hope this helps.

**Comment by chrisfillmore** [comment]
Thanks Joey, sorry I haven't looked at this again yet. I will submit a PR early next week so I don't fall too far behind.

**Comment by joeyparrish** [comment]
Sounds good.  We are no rush at the moment.  It's something we plan to give more attention to for v2.4 in January.

**Comment by chrisfillmore** [comment]
I invite anyone that has HLS-FP content that they can test with to give this commit a shot. I will need to integrate this into my custom app to do more thorough testing. That probably won't happen till January. Merry Christmas everyone!

**Comment by fikiland** [comment]
you can convert mp4 to a fairplay using bento4 here, 
https://www.bento4.com/developers/hls/
and we can do it in AWS using CLI

but sorry i can't share my link due to term with my company

but still we have problem.

i have play fairplay using shaka player, but it only work for 15% of all video converted, i dunno why but it give error "
![fairplay error](https://user-images.githubusercontent.com/1190374/34297965-c18039c0-e74d-11e7-8f85-3c3390b9e7fe.jpg)


invalidAccessError (DOM Exception 15): The object does not support the operation or argument.
Failed to load resource: the server responded with a status of 403(License denied. the license request contains a key ID that is not present in the entitlement.) 

i try repackage it again but still it doesn't work
but in the morning it works like a charm, maybe cache in AWS?
no idea

but still the other video got same error. that is my experience and problem. anyone got same problem and can tackling it? please help 
"

**Comment by fikiland** [comment]
sorry my issue already close, it just miss the id between player to axinom and the database in the backend

btw since safari doesn't return any eme 
i use user_agent to define the eme,
and we just give fairplay to mac os user

**Comment by jesteross** [comment]
Hi All

Thanks all contributors for great work!

Could you please say, is it planned FairPlay support?
And maybe some approximate date?

**Comment by joeyparrish** [comment]
@sridhard, no updates yet.

You are mistaken about #1447, though.  It does not say fps1.0 is supported.  Rather, it says we plan to support it, and refers back to this issue to track progress.

I know this issue is important to many people, and I regret the delay in implementation.  I promise that we will give updates as soon as there are updates to give.  Thank you for your continued patience.

**Comment by htleeab** [comment]
@chrisfillmore My company is trying to play Hls-fps. The hls streams of my company have not pssh boxes and webkitneedkey return 149 bytes data. I extracted the key id(=content id) from playlist to simulate the initData like fps.1_0. License server can return CKC message but the video cannot play. 

I'm wondering the initData of fps.2_0 is different from fps.1_0. Could you please tell me what is the correct value and pattern of initData from webkitneedkey? or is there any workable sample for fps.2_0 or fps.3_0? 

Thank you.

**Comment by chrisfillmore** [comment]
Yes the initData is different from fps1.0 to fps2.0. In 1.0, the initData is the SKD value from the manifest. In 2.0 it is the PSSH (as far as I know). I'm not aware of a working demo for fps2.0.

**Comment by htleeab** [comment]
@chrisfillmore 
Thank for the reply.
My colleague generated a hls + fairplay stream with pssh box. 
The initData return from webkitKeyMessage is array of char code of stringified json of sinf box
`{"sinf" : [    "AAAADGZybWFtcDRhAAAAFHNjaG0AAAAAY2JjcwABAAAAAAA5c2NoaQAAADF0ZW5jAQAAAAAAAQDJyuv\/nVBJRoJ1xpgf1uGcEDBBHQdj5UTQu8Xqnj8si+U="  ]}`
the value of sinf is 64 encoded sinf box. 
Safari can handle both above sinf box and the initData of fps.1_0 ( string of skd url with length padded to multiple of 16 bytes).

When load source with video and audio stream, one of the key will be added and the another one will throw webkitkeyerror, no matter if the keys are same or not. Also, the session ids are empty. 
But when there is only one key session (video-only or audio-only), it can play successfully. :D
My colleague think it may be because the method safari handle the media source. Safari native player only trigger one webkitneedkey event when the keys are same. But most of MSE player with DRM require two sessions.
Apart from that, my colleague generated a CENC fps stream, it cannot be played but it create session with id.
It is related to the data of pssh box? Currently we using system_id = [29 70 1f e4 3c c7 4a 34 8c 5b ae 90 c7 43 9a 47]

**Comment by htleeab** [comment]
I fount out that the fps 1 license server I using get the key id outside of SPC without verify with the content of SPC. That maybe the reason that I can get the correct key back and play CBCS video/audio only. The asset id decrypted by license server is the whole input of create session. If I pass the initData as-is, the license server receive the whole sinf box.

I tested a stream with same key for video and audio. I process the sinf box returned from webkitneedkey and extract key id to create session so that there is only one session and safari generates SPC with key id that match the fps1 license server I using. The key added correctly without error, but the video is stalling and cannot play.

Maybe the system id only for CENC(CTR) and the data stored in the pssh box of init segment have some impact. I cannot test CENC stream since it requires IV with different length and it requires to modify the license server.

**Comment by fadomire** [comment]
dont mean to put any pressure (as if i could) but is Fairplay DRM support still planned for 2.5 ?
i know it is blocked by #816 

if i can be of any help to make Fairplay DRM work just let me know

**Comment by htleeab** [comment]
Safari 12.1 have some new webkit features.

> Modern Encrypted Media Extensions API
> WebKit has an updated implementation of the Encrypted Media Extensions (EME) API for providing encrypted video or audio content. This updated version also no longer uses a webkit- prefix. The modern EME implementation includes a number of web developer features and conveniences not found in the original, prefixed API. Developers are encouraged to use feature detection to determine if the prefixed or the non-prefixed version of the API is available to ensure compatibility with older versions.
> 
> See the Encrypted Media Extensions standards specifications for more information.
https://webkit.org/blog/8718/new-webkit-features-in-safari-12-1/

I test some API with setting video.src=*.m3u8.
Safari 12.1 trigger 'webkitneedkey' and 'encrypted'.
Safari 12.1 still support previous fairplay flow (webkit-prefixed v0.1b API?).
Safari 12.1  have window.MediaKeys and video.setMediaKeys.
Safari 12.1 have Navigator.requestMediaKeySystemAccess but return MediaKeySystemAccess of clearkey only.
Do anyone know the EME Spec version of the non-prefixed API and if it works for fairplay?

**Comment by joeyparrish** [comment]
I'm using Safari 12.1 (13607.1.40.1.5) and I don't see any unprefixed EME APIs available (setMediaKeys, MediaKeys, requestMediaKeySystemAccess, etc.)

So for now, I'm working on the prefixed APIs (WebKitMediaKeys, webkitSetMediaKeys, etc).

**Comment by joeyparrish** [comment]
If anyone monitoring this issue has FPS sample content they'd like us to check, to make sure our new implementation is compatible, please send m3u8 URI **and** license server URI **and** certificate URI.  You can send them to shaka-player-issues@google.com if you don't want to post them publicly.  Thanks!

**Comment by htleeab** [comment]
> I'm using Safari 12.1 (13607.1.40.1.5) and I don't see any unprefixed EME APIs available (setMediaKeys, MediaKeys, requestMediaKeySystemAccess, etc.)
> 
> So for now, I'm working on the prefixed APIs (WebKitMediaKeys, webkitSetMediaKeys, etc).

Thanks for testing :)
I'm sorry that I forgot to mention that the previous testing result is tested on macbook. I'm not sure if mac or other devices return same result. 
Beside, iPad Pro (10.5-inch, software version 12.2) also trigger encrypted and have setMediaKeys and MediaKeys. But requestMediaKeySystemAccess reject all key systems, including clearkey.

**Comment by joeyparrish** [comment]
I'm also on a Mac.  I'm not sure what the difference is.  Is there a setting in the Develop menu for this?

**Comment by avelad** [comment]
@joeyparrish https://webkit.org/blog/8718/new-webkit-features-in-safari-12-1/

![eme](https://user-images.githubusercontent.com/8983024/56729141-d5990580-6754-11e9-9ce7-87dcc32c7d70.png)

**Comment by joeyparrish** [comment]
@avelad, yes, I see that, but the fact remains that window.MediaKeys is undefined in Safari 12.1 on my Mac.  I'm not running Mojave, though.

Can someone with Mojave confirm or deny that window.MediaKeys shows up in the JS debugger on Safari 12.1?  (On a blank page, not Shaka Player.  Our polyfills will create a fake one regardless.)

**Comment by senaev** [comment]
> Can someone with Mojave confirm or deny that window.MediaKeys shows up in the JS debugger on Safari 12.1? 

yes, there is MediaKeys, but i did not find the way to detect fireplay anyway

Now we are use WebKitMediaKeys.isTypeSupported, and we are in search of unprefixed api to.

<img width="1348" alt="Google 2019-04-25 23-19-33" src="https://user-
images.githubusercontent.com/7939529/56765835-a6b07d00-67b0-11e9-887e-631a23c90c37.png">

**Comment by joeyparrish** [comment]
`isTypeSupported` is not a part of the MediaKeys API.  Instead, you are supposed to use `navigator.requestMediaKeySystemAccess`, which is much more complicated.  Thankfully, Shaka Player handles that API for you.

Since this unprefixed API is dependent on a specific OS version, I think it's best if I continue developing the polyfill to support older OS versions.  Then I'll perform the update to Mojave and test again.  :-)

**Comment by senaev** [comment]
> `isTypeSupported` is not a part of the MediaKeys API. Instead, you are supposed to use `navigator.requestMediaKeySystemAccess`, which is much more complicated. Thankfully, Shaka Player handles that API for you.

Yeah, used to try, but there is also no way to detect fireplay with `navigator.requestMediaKeySystemAccess` api on 12.1 version. If you know how to do it, share it please!

**Comment by joeyparrish** [comment]
If that's the case, we'll be stuck using prefixed until Apple implements the whole thing.  I'll find out soon.  Probably tomorrow.

**Comment by joeyparrish** [comment]
A teammate has confirmed that requestMediaKeySystemAccess is available on Safari 12.1 on Mojave specifically.  It's not on 12.1 on High Sierra.

**Comment by joeyparrish** [comment]
Thank you @avelad for the content.  My changes are still in review, but I have your content working with Apple's HLS and `fps.1_0`!

If anyone else has sample content to contribute, please send them our way!

**Comment by osmestad** [comment]
I sent a link with some TIDAL sample content to shaka-player-issues@google.com now, would be great if you can test that!

**Comment by joeyparrish** [comment]
Will do!  Thank you!

**Comment by avelad** [comment]
@joeyparrish Can you update README.md file with current Fairplay status?

**Comment by joeyparrish** [comment]
Yes, thanks for reminding me!

**Comment by mathieuedet** [comment]
I'm trying to implement Fairplay with new shaka-player version (2.5) ? Can you update documentation to include configuration for Fairplay ?

**Comment by joeyparrish** [comment]
Sure.  In the mean time, here's what you need to know.

You use FairPlay just like any other DRM in terms of the license server config:

```js
player.configure({
  drm: {
    servers: {
      'com.apple.fps.1_0': '/path/to/my/server',
    },
  },
});

// or

player.configure('drm.servers.com\\.apple\\.fps\\.1_0', '/path/to/my/server');
```

In addition, FairPlay requires you to provide a certificate through the "advanced" DRM config:

```js
const response = await fetch('/path/to/my/cert.der');
if (!response.ok) {
  // handle error
}
const cert = await response.arrayBuffer();

player.configure({
  drm: {
    advanced: {
      'com.apple.fps.1_0': cert,
    },
  },
});

// or

player.configure('drm.advanced.com\\.apple\\.fps\\.1_0', cert);
```

If your license server doesn't use the default formatting for FairPlay license requests/responses, you can use request and response filters to transform them.  See the "License Wrapping" tutorials for generic examples of that.

I hope that helps you get started.  We'll work on improving the docs around FairPlay so that these things are clearer.

**Comment by sandeep-mirpuri** [comment]
@joeyparrish I'm so excited about this. Been waiting for it for ages :). But I can't get it to work. I am setting the licenseUrl and certificate ArrayBuffer as described in your comment above. Then trying to load the corresponding m3u8 manifest. But I get a couple of different errors.

1) ` invalid config, wrong type for .drm.advanced.com.apple.fps.1_0`
2) `Error code: 6010 - Shaka Error DRM.ENCRYPTED_CONTENT_WITHOUT_DRM_INFO`

I assumed that it might be an issue with getting the license key since we need to append a session key and ticket number to the request. I tried registering a request filter for this, but it seems its not even getting as far as the license, since the filter doesn't seem to be firing at all (fires fine on chrome using widevine). This is the line for the key info in the manifest.

`#EXT-X-SESSION-KEY:METHOD=SAMPLE-AES,URI="skd://key-server-url",KEYFORMAT="com.apple.streamingkeydelivery",KEYFORMATVERSIONS="1"`

I have also attached an image with the logs in my safari browser from shaka

Not sure what is happening here. I am working on getting permissions to share the manifest with you but until then, any pointers? (At the moment we are using a raw implementation using safari's webkit apis. This implementation works fine. Let me know if steps describing this implementation are of any use)

<img width="714" alt="Screen Shot 2019-05-23 at 2 22 27 pm" src="https://user-images.githubusercontent.com/30884334/58225726-d084a180-7d66-11e9-8fa9-520a29805369.png">
